### ISSUE TITLE
Use caption to suggest depictions

### ISSUE THREAD

**Issue by nicolas-raoul** [commons-app/apps-android-commons#5422]
If I caption my picture "Papilio machaon on Asteraceae flower in Croatia", then the ap should ideally suggest these depictions:

- papilio machaon
- asteraceae
- flower
- Croatia

First step: Parse the caption into expression could be asked to an LLM if one is available locally on the device. That means only on Pixel 8 Pro for now it seems, and we need to identify the API: https://www.theverge.com/2023/12/6/23989591/google-gemini-ai-model-pixel-8-pro-recorder-smart-reply
If no such technology is available, use some more traditional tokenization, maybe even split on each space character.

Second step: Search for 1 or 2 Wikidata items that match each expression found in the first step.

Third step: Add these suggestions to the list of all suggestions.

**Comment by rohit9625** [comment]
I don't think splitting the caption would work, because how do we identify which words need to be shown in depictions?

**Comment by nicolas-raoul** [comment]
@rohit9625 It is pretty easy using LLM:
![Screenshot_20240109-075634.png](https://github.com/commons-app/apps-android-commons/assets/99590/c8fb07af-9a2b-4bd1-ad07-004504c6e358)

Even splitting word-by-word would work reasonably well.

Prompt for further improvements:
```
You are a syntax parser that understands all human languages. For each input sentence, extract all nouns from the sentence. Include each adjective with its noun. Do not output anything but the list of nouns.

Example: "Papilio machaon on Asteraceae flower in Croatia"
 - papilio machaon
 - asteraceae
 - flower
 - Croatia

Now do the same for: "Oxcart passing on a wooden bridge in Papua New Guinea"
```

**Comment by RitikaPahwa4444** [comment]
> That means only on Pixel 8 Pro for now it seems

Is on-device execution necessary for this use case? The API is available for [target API level 21 and higher,](https://ai.google.dev/tutorials/android_quickstart#prerequisites) and should work if we're not going for Gemini Nano.

**Comment by nicolas-raoul** [comment]
> Is on-device execution necessary for this use case?

On-device execution is necessary. As a privacy policy, we do not send any HTTP request to any server outside of Wikimedia (_and Mapbox until recently but hopefully this should be resolved soon_).

Details: https://github.com/commons-app/commons-app-documentation/blob/master/android/Privacy-policy.md#privacy-policy

**Comment by RitikaPahwa4444** [comment]
Thank you for clarifying! Since captions are publicaly available, I'd not thought of it from this perspective. [This](https://codelabs.developers.google.com/kerasnlp-tflite#0) seems like an on-device LLM, but I'm still exploring😅.

**Comment by nicolas-raoul** [comment]
Great link thanks Ritika! I wonder how many megabytes that would add to the app... if not too many that's definitely an option.

**Comment by kanahia1** [comment]
Hey @nicolas-raoul, I have found a model which we utilize to achieve NER [https://huggingface.co/dslim/bert-base-NER](https://huggingface.co/dslim/bert-base-NER). They have provided ONNX model which we can to convert to .tflite model, then can do the implementation in the app.

I tested it with `Papilio machaon on Asteraceae flower in Croatia` these were the results 

```
[
  {
    "entity_group": "PER",
    "score": 0.4467669427394867,
    "word": "Pa",
    "start": 0,
    "end": 2
  },
  {
    "entity_group": "PER",
    "score": 0.8728312849998474,
    "word": "mac",
    "start": 8,
    "end": 11
  },
  {
    "entity_group": "MISC",
    "score": 0.9185755848884583,
    "word": "Asteraceae",
    "start": 19,
    "end": 29
  },
  {
    "entity_group": "LOC",
    "score": 0.9997634291648865,
    "word": "Croatia",
    "start": 40,
    "end": 47
  }
]
```

I believe accuracy is good as it provides us with `Asteraceae` and `Croatia`

**Comment by kanahia1** [comment]
@nicolas-raoul, I have converted the model to .tflite ~ 411 MB which is very large when we have to use in Android App.

Is it possible we can use MLKit by Google in the App ?
https://github.com/googlesamples/mlkit/tree/master/android/entityextraction

**Comment by nicolas-raoul** [comment]
@kanahia1 Great research, thanks!

I feel like 50MB should be maximum...

MLKit sounds great! Can it extract keywords? How big is it? 🙂

**Comment by kanahia1** [comment]
@nicolas-raoul Sure, It can extract 
https://github.com/googlesamples/mlkit/tree/master/android/entityextraction?rgh-link-date=2024-03-04T06%3A00%3A08Z
Entity extraction has an app size impact of up to `~5.6MB.`

They have provided examples - 

Input Text - `Meet me at 1600 Amphitheatre Parkway, Mountain View, CA, 94043 Let’s organize a meeting to discuss.`
Output Text - `1600 Ampitheatre Parkway, Mountain View, CA 94043`

More examples are available [here](https://developers.google.com/ml-kit/language/entity-extraction)

**Comment by kanahia1** [comment]
@nicolas-raoul 
I tested MLKit, results I found were not satisfactory. I think dropping the idea of using MLKit would be better until the Google make it more effective since it also have multi-language support.

```
Test 1 
Input: Dust storm approaching Stratford, Texas
Output: There are no entities detected

Test 2
Input: Aerial view of the city of Mayfield on December 12.
Output: December 12.

Test 3
Input: Interior view of the main nave of the Cathedral in Narbonne, France. The Roman Catholic church is dedicated to Saints Justus and Pastor and it begun its construction in 1272. The choir was finished in 1332, but the rest of the gothic building was never completed, as the result of many factors including sudden changes in the economic status of Narbonne, its unusual size and geographical location (to complete it would have meant demolishing the city wall) and financial constraints.
Output: There are no entities detected

Test 4
Input: Mt Technical from Lewis tops, Lewis Pass Scenic Reserve, New Zealand
Output: There are no entities detected
```

If I get enough time by end of this month, I will try to train a custom model 🙂

**Comment by Thejas775** [comment]
> If I caption my picture "Papilio machaon on Asteraceae flower in Croatia", then the app should ideally suggest these depictions:
> 
> * papilio machaon
> * asteraceae
> * flower
> * Croatia
> 
> First step: Parse the caption into expressions could be asked to an LLM if one is available locally on the device. That means only on Pixel 8 Pro (or above) and Samsung Galaxy S24 (or above) for now but hopefully more makers will follow. On devices where no such technology is available, the app can fall back to some more traditional tokenization, maybe even split on each space character, or just skip entirely.
> 
> Second step: Search for 1 or 2 Wikidata items that match each expression found in the first step.
> 
> Third step: Add these suggestions to the list of all suggestions.

We can try running Gemini Nano using the AICore but it would support very few devices.
Well I also found this : [](https://ai.google.dev/edge/mediapipe/solutions/genai/llm_inference/android)
Maybe this will work.

**Comment by nicolas-raoul** [comment]
> AICore but it would support very few devices

Yes, this project is very long-term (and not urgent) so even if it currently supports only few devices that's OK. Also, development is possible without owning such a device (_we will use stubs, and someone else who has such a device will test for real_).

> Well I also found this : https://ai.google.dev/edge/mediapipe/solutions/genai/llm_inference/android

Fantastic! Since it could be used for more things than just depictions, I mentioned it here: https://github.com/commons-app/apps-android-commons/issues/6143#issuecomment-2609014967
